Make clusters geodistributed by default.

m-schneider opened the pull request:
Roachprod will start up a new cluster with 2 nodes in US east, 1 in US west and one
in Europe by default.

Adds zones flag so that the default can be overridden.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachlabs/roachprod/19)
<!-- Reviewable:end -->

mberhault commented on gcloud.go:
```diff
@@ -106,55 +142,60 @@ func createVMs(names []string, opts VMOpts) error {
+		args = append(args, "--zone", zones[i%len((zones))])
+		//args = append(args, "--additional-zones",strings.Join(zones[1:],","))
+		args = append(args, name)
+		cmd := exec.Command("gcloud", args...)
```
this is going to be much slower. How about grouping the commands by zone and providing the list of all VMs in that zone? An alternative would be to run commands in goroutines, but what has a risk of making too many gcloud calls.

benesch replied:
I vote goroutines; I think the common case (to start) will be spinning up _n_ nodes in _n_ regions.

mberhault replied:
Well, the common case for geo-distributed. But if you're running all nodes in one region, why not let `gcloud` handle that for you? I haven't tried 50 concurrent instances of gcloud, but azure would melt your computer.

mberhault commented on gcloud.go:
```diff
@@ -9,15 +9,42 @@ import (
+	return nil
+}
+
+var zones = ZoneList{"us-east1-b", "us-west1-b", "europe-west2-b", "us-east1-b"}
```
multi-zone definitely shouldn't be the default.

danhhz replied:
I actually disagree pretty strongly. Global data is a very important story of the 2.0 release and we need to be running geo-distributed clusters as much as possible to catch issues. If someone _really_ needs a single dc cluster, then they can use the flag to opt out.

mberhault replied:
Sure, but right now everyone's workflow it to do performance work that does not involve geo-distributed cluster. If you want to make it easy to have zones, use a `--geo` bool flag that automatically picks among the (flag default) zones. (eg: round-robin through the zones).
You don't be breaking anyone's existing use, and you still have a simple `--geo` invocation.

petermattis replied:
While I agree with the importance of testing geo-distributed clusters, please don't move my cheese. More specifically, changing the default will break nightly usage of roachprod for perf tests, perhaps jepsen tests (not sure if it is used there yet), perhaps other stuff. I'm sure we have more important work to do then fixing up those problems and educating engineers about this change.

danhhz replied:
We have more important work to do than increasing our testing of the headline theme of the 2.0 release? I'm sorry but I disagree. And educating engineers about the flag to turn it off (which at the moment is probably only the people doing single-dc performance work) seems like a pretty low investment.

I'm happy to fix up nightly perf tests (though arguably we should also be running the perf tests on a geo-distributed cluster as well). Same for Jepsen, though once the performance is good enough, seems like that should be tested geo-distributed.

That said, I think there does need to be a `--single-dc` option to disable it instead of the `--zones=us-east1-b,us-east1-b,us-east1-b,us-east1-b` you'd have to do with the PR as it currently stands.

mberhault replied:
There are plenty of features and performance work being done that does **not** want geo-distributed clusters. We can switch to making it the default in the future, but especially when adding something which will most likely break on setup/GCE quotas/random other reasons, it's better to have it off by default.

You can make it simple by doing something along the lines of:
```
--zones=us-east1,us-west, europe
--geo=false
```

When `--geo` is false, all N nodes are created in the first zone specified in `--zones`.
When `--geo` is true, each zone gets `num_nodes / num_zones` nodes (either round-robin or 1-3 in zone 1, 4-6 in zone 2, 7-8 in zone 3).

This makes flipping the `--geo` flag the only thing necessary unless you want to also configure exactly which zones you want. Everyone currently using roachprod will keep using it as it. Anyone who wants geo-distributed clusters just sets `--geo`.

petermattis replied:
> We have more important work to do than increasing our testing of the headline theme of the 2.0 release?

I might argue that some of the stability work on non-distributed clusters is as or more important. Regardless of the priority, I'm pretty sure you won't increase testing of geo-distributed clusters by forcing those clusters on users of `roachprod` by default. I'm massively in agreement that we should increase testing of geo-distributed clusters. And I fully support making it easy to create such clusters. But I'm confused about the insistence on changing the default for an existing tool. Seems a step too far in my opinion.

I'm going to stop arguing about this, though and leave Marc to carry the torch. I'll adjust if the default changes.

danhhz replied:
I like your proposed flags a lot. And I don't want to hold up this PR on the default. Let's do your proposal, but do expect me to bring up revisiting the default soon. @m-schneider sgty?

mberhault replied:
I'm happy revisiting. Even without switching the default we should look at which of our automated tests should set `--geo` pretty might right after the PR is merged.

tschottdorf replied:
Not planning to get pulled into this discussion too much, but I basically agree with @danhhz (and will let him carry the 🔦) and wish we'd just make this the default and adjust the 2-3 places that don't want it.

> Regardless of the priority, I'm pretty sure you won't increase testing of geo-distributed clusters by forcing those clusters on users of roachprod by default. 

The long list of problems discovered by @dianasaur323 on the PM cluster suggests otherwise.

petermattis replied:
> The long list of problems discovered by @dianasaur323 on the PM cluster suggests otherwise.

I wasn't suggesting there aren't problems with geo-distributed clusters, just that changing the default for `roachprod` won't increase testing of such clusters. What will increase the testing of geo-distributed clusters is making it trivially easy to create such clusters, which this PR accomplished both with and without a change to the default.

tschottdorf replied:
> I wasn't suggesting there aren't problems with geo-distributed clusters, just that changing the default for roachprod won't increase testing of such clusters.

🤷‍♂️ I guess we're going to have to agree to disagree. More use = more testing and the most versatile testing we can get is everyone using geo-distributed clusters unless they know for a fact they care about the local commit latencies. If you find yourself wanting to create non-distributed clusters all the time, I wouldn't be opposed to sticking an env var in there that you can set in your `.bashrc` so that you get the default behavior you want. For anyone else who doesn't care, I can only see benefits to giving them a cluster that's at least somewhat indicative of the feature we want to have polished by 2.0. And that includes performance work too - I doubt there's an innate drive in many of us to get performance numbers of geo-distributed clusters, but that's something to look at as well.

mberhault replied:
Can we please stop? We're not suggesting we **never** make geo-distributed setups the default, we're just saying not right away. May main reasons are:
* this is brand new behavior, keeping the existing one should be the default
* it's going to fail in fun and interesting ways, that's what cloud stuff does. Let's find and fix those first.

So I would propose the following sequence of events (without any fixed timeline, but "soon" comes to mind):
1. merge PR with default of `--geo=false`
1. find automated tests that use roachprod. For all of them, determine if they absolutely need single-region testing and set `--geo=false` for those. Set `--geo=true` progressively on others
1. get everyone to use the latest roachprod. Document and encourage use of geo-distributed testing.
1. flip the default to true and tell everyone.

Once the last is done and you still want to run in a single region, then you'll know about `--geo`.

If you want to make the timeline as short as possible, exercise `--geo=true` as much as you can, find issues, troubleshoot them, and fix them.

mberhault commented on gcloud.go:
```diff
@@ -106,55 +142,60 @@ func createVMs(names []string, opts VMOpts) error {
-		"compute", "instances", "delete",
-		"--delete-disks", "all",
-	}
+func deleteVMs(names []string, vmZones []string) error {
```
this should probably take a `VM` struct now that we need more than the name.

mberhault commented on gcloud.go:
```diff
@@ -9,15 +9,42 @@ import (
+// interface.
+type ZoneList []string
+
+var _ pflag.Value = &ZoneList{}
```
this should be with all the other flags in `main.go`

mberhault commented on gcloud.go:
```diff
@@ -106,55 +142,60 @@ func createVMs(names []string, opts VMOpts) error {
+		}
 	}
 	return nil
 }
```
what happened to `extendVM`? That went in over a month ago.

BramGruneir commented:
Just wanted to leave a comment here to just highlight that what we call `localities` and `zones` differ from what gce calls `zones`.  And of course, we allow more than just top level localities, which may be important.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-:-L0KqJyIGq__GqUTj9Gh:bebwlqt)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Sorry Marc, didn't intend to gate the fate of this pr on the default. Your
suggestions LGTM!

On Thu, Dec 14, 2017, 6:19 PM marc <notifications@github.com> wrote:

> *@mberhault* commented on this pull request.
> ------------------------------
>
> In gcloud.go
> <https://github.com/cockroachlabs/roachprod/pull/19#discussion_r157088729>
> :
>
> > +}
> +
> +// Type returns the underlying type in string form. This is part of pflag's
> +// value interface.
> +func (*ZoneList) Type() string {
> +	return "ZoneList"
> +}
> +
> +// Set adds a new value to the ZoneList. It is the important part of
> +// pflag's value interface.
> +func (zoneList *ZoneList) Set(value string) error {
> +	*zoneList = ZoneList(strings.Split(value, ","))
> +	return nil
> +}
> +
> +var zones = ZoneList{"us-east1-b", "us-west1-b", "europe-west2-b", "us-east1-b"}
>
> Can we please stop? We're not suggesting we *never* make geo-distributed
> setups the default, we're just saying not right away. May main reasons are:
>
>    - this is brand new behavior, keeping the existing one should be the
>    default
>    - it's going to fail in fun and interesting ways, that's what cloud
>    stuff does. Let's find and fix those first.
>
> So I would propose the following sequence of events (without any fixed
> timeline, but "soon" comes to mind):
>
>    1. merge PR with default of --geo=false
>    2. find automated tests that use roachprod. For all of them, determine
>    if they absolutely need single-region testing and set --geo=false for
>    those. Set --geo=true progressively on others
>    3. get everyone to use the latest roachprod. Document and encourage
>    use of geo-distributed testing.
>    4. flip the default to true and tell everyone.
>
> Once the last is done and you still want to run in a single region, then
> you'll know about --geo.
>
> If you want to make the timeline as short as possible, exercise --geo=true
> as much as you can, find issues, troubleshoot them, and fix them.
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachlabs/roachprod/pull/19#discussion_r157088729>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135PYmQA_dYDsU6sO6i4J_GKIDjbagks5tAa0BgaJpZM4RBMz7>
> .
>
-- 

-- Tobias

jordanlewis commented:
BTW I tried this patch out - it's great that it made it so easy to make a distributed cluster! Unfortunately it seems to have broken roachprod destroy:

```
[16:24:1]% roachprod destroy cc-test
Destroying cluster jordan-cc-test with 4 nodes
Error: Command: gcloud [compute instances delete --delete-disks all --project cockroach-ephemeral --zone compute jordan-cc-test-0001]
Output: The following instances will be deleted. Any attached disks configured
 to be auto-deleted will be deleted unless they are attached to any
other instances or the `--keep-disks` flag is given and specifies them
 for keeping. Deleting a disk is irreversible and any data on the disk
 will be lost.
 - [jordan-cc-test-0001] in [compute]

Do you want to continue (Y/n)?
ERROR: (gcloud.compute.instances.delete) Failed to fetch some instances:
 - Invalid value for field 'zone': 'compute'. Unknown zone.

: exit status 1
Usage:
  roachprod destroy <cluster ID> [flags]

Flags:
  -u, --username string   Username to run under, detect if blank

Command: gcloud [compute instances delete --delete-disks all --project cockroach-ephemeral --zone compute jordan-cc-test-0001]
Output: The following instances will be deleted. Any attached disks configured
 to be auto-deleted will be deleted unless they are attached to any
other instances or the `--keep-disks` flag is given and specifies them
 for keeping. Deleting a disk is irreversible and any data on the disk
 will be lost.
 - [jordan-cc-test-0001] in [compute]

Do you want to continue (Y/n)?
ERROR: (gcloud.compute.instances.delete) Failed to fetch some instances:
 - Invalid value for field 'zone': 'compute'. Unknown zone.

: exit status 1
```

mberhault commented:
Yeah, this isn't ready yet, all commands were using a fixed zone before, but most gcloud commands (other than list) are for a single zone so you can't just throw all machines at it.
When you're done with your cluster, can you destroy them the old fashioned way? (through the google UI)

m-schneider commented:
Jordan, should be fixed now if you want to try it, destroy should kill a cluster started without this patch.

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[gcloud.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L0Glwp3UvXRoZsKoYx9-r1-25:-L0fc_SeCUmdDWqHTn9n:b-896fix) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c2061be31237664d8a429b75674621d46a9aa140/gcloud.go#L25)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

this should be with all the other flags in `main.go`
</blockquote></details>

Done.

---

*[gcloud.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L0Glwp3UvXRoZsKoYx9-r1-46:-L0KwVGMYL2BTimX5G6b:bt3tje5) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c2061be31237664d8a429b75674621d46a9aa140/gcloud.go#L46)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Can we please stop? We're not suggesting we **never** make geo-distributed setups the default, we're just saying not right away. My main reasons are:
* this is brand new behavior, keeping the existing one should be the default
* it's going to fail in fun and interesting ways, that's what cloud stuff does. Let's find and fix those first.

So I would propose the following sequence of events (without any fixed timeline, but "soon" comes to mind):
1. merge PR with default of `--geo=false`
1. find automated tests that use roachprod. For all of them, determine if they absolutely need single-region testing and set `--geo=false` for those. Set `--geo=true` progressively on others
1. get everyone to use the latest roachprod. Document and encourage use of geo-distributed testing.
1. flip the default to true and tell everyone.

Once the last is done and you still want to run in a single region, then you'll know about `--geo`.

If you want to make the timeline as short as possible, exercise `--geo=true` as much as you can, find issues, troubleshoot them, and fix them.
</blockquote></details>

Added the geo flag and set it to false by default. When false it'll just use the first region in the zones list to set up the cluster. I also go rid of the round robin algorithm to assign zone names, they'll now be continuous for a region.

---

*[gcloud.go, line 171 at r1](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L0Glwp3UvXRoZsKoYx9-r1-171:-L0fbliPmMWR_J7Mdb8K:bjqnyz7) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c2061be31237664d8a429b75674621d46a9aa140/gcloud.go#L171)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Well, the common case for geo-distributed. But if you're running all nodes in one region, why not let `gcloud` handle that for you? I haven't tried 50 concurrent instances of gcloud, but azure would melt your computer.
</blockquote></details>

Added as a goroutine per zone.

---

*[gcloud.go, line 182 at r1](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L0Glwp3UvXRoZsKoYx9-r1-182:-L0fbnRdL3LIg369K_Bn:b-x1di1s) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c2061be31237664d8a429b75674621d46a9aa140/gcloud.go#L182)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

this should probably take a `VM` struct now that we need more than the name.
</blockquote></details>

I looked into it and I think the reason that deleteVMS previously worked on a list of names vs VMs is that this gets called from both monitor.go and cluster.go which are operating on different base structures. Monitor.go has a a JsonVmlist and cluster.go uses a VMList.

---

*[gcloud.go, line 201 at r1](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L0Glwp3UvXRoZsKoYx9-r1-201:-L0fcYtsh2IaIEJ-2AMj:b-896fix) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c2061be31237664d8a429b75674621d46a9aa140/gcloud.go#L201)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

what happened to `extendVM`? That went in over a month ago.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-:-L0fb-Uk0feP_Cl80XVr:b-h08b7t)*
<!-- Sent from Reviewable.io -->

mberhault commented on gcloud.go:
```diff
@@ -106,60 +117,105 @@ func createVMs(names []string, opts VMOpts) error {
-	args = append(args, "--labels", fmt.Sprintf("lifetime=%s", opts.Lifetime))
+	totalNodes := float64(len(names))
+	totalZones := float64(len(zones))
+	nodesPerRegion := int(math.Ceil(totalNodes / totalZones))
```
nit: s/nodesPerRegion/nodesPerZone/`.

mberhault commented on gcloud.go:
```diff
@@ -106,60 +117,105 @@ func createVMs(names []string, opts VMOpts) error {
+	var wg sync.WaitGroup
+	wg.Add(len(zones))
+
+	for i < len(names) {
```
This loop's logic deserves a comment.

mberhault commented on gcloud.go:
```diff
@@ -106,60 +117,105 @@ func createVMs(names []string, opts VMOpts) error {
+
+	for i < len(names) {
+		// Fixed args.
+		args := []string{
```
you can keep the common args outside the loop as long as you use a new per-iteration arg. eg:
```
args := []string{ ....}
for .... {
  zoneArgs = append(args, newArgs...)
  run(gcloud, zoneArgs)
}
```

mberhault commented on monitor.go:
```diff
@@ -78,7 +78,7 @@ func monitorClusters(cloud *Cloud, filename string, destroyAfter time.Duration)
 	maybeAddVMs(cloud.BadNetwork)
 	if len(badVMs) > 0 {
-		if err := deleteVMs(badVMs.Names()); err != nil {
+		if err := deleteVMs(badVMs.Names(), badVMs.Zones()); err != nil {
```
You'll need to update the roachprod binary on the shared machine. See https://github.com/cockroachlabs/production/tree/master/shared
Just stick a linux binary in the binaries subdirectory, then `crl-put shared 1  binaries/roachprod roachprod`. The cron job will just pick it up.

tschottdorf commented:
Could you update the commit message and PR description?

LGTM otherwise with the open comments addressed and @mberhault happy.

---

Reviewed 3 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[cluster.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L1wKnt1xmDRTUToLzuU:-L1wKnt1xmDRTUToLzuV:b-jsc5fe) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c865d2a28b7f8411fb82df35404681838a48b9bb/cluster.go#L127)):*
> ```Go
> 		publicIP := vm.NetworkInterfaces[0].AccessConfigs[0].NatIP
> 		vmZones := strings.Split(vm.Zone, "/")
> 		zone := vmZones[len(vmZones)-1]
> ```

Add a comment here. It's unclear to me why the last one is the right one.

---

*[gcloud.go, line 208 at r3](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L1wMbGyvpXgV1YZygoD:-L1wMbGyvpXgV1YZygoE:b-kvh7pd) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c865d2a28b7f8411fb82df35404681838a48b9bb/gcloud.go#L208)):*
> ```Go
> 			output, err := cmd.CombinedOutput()
> 			if err != nil {
> 				return errors.Wrapf(err, "Command: gcloud %s\nOutput: %s", args, output)
> ```

This is silently swallowed. You could use an `errgroup`.

---

*[main.go, line 310 at r3](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L1wMpoY-aFcImv0OAD9:-L1wMpoY-aFcImv0OADA:b-aouoij) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c865d2a28b7f8411fb82df35404681838a48b9bb/main.go#L310)):*
> ```Go
> 	createCmd.Flags().IntVarP(&numNodes, "nodes", "n", 4, "Number of nodes")
> 	createCmd.Flags().StringVarP(&username, "username", "u", "", "Username to run under, detect if blank")
> 	createCmd.Flags().VarP(&zones, "zones", "z", "Zones for cluster.")
> ```

No dot here and below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-:-L1wKf_7JwfkQDyZ4wr0:b-u1t0fc)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 1 of 4 files reviewed at latest revision, 12 unresolved discussions.

---

*[cluster.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L1wKnt1xmDRTUToLzuU:-L1xAkcBl7cHlFrCBRHO:b-896fix) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c865d2a28b7f8411fb82df35404681838a48b9bb/cluster.go#L127)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Add a comment here. It's unclear to me why the last one is the right one.
</blockquote></details>

Done.

---

*[gcloud.go, line 126 at r3](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L0Glwp3UvXRoZsKoYx9-r3-126:-L1xAtB6ZPbBKCYoXMmX:b-896fix) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c865d2a28b7f8411fb82df35404681838a48b9bb/gcloud.go#L126)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

nit: s/nodesPerRegion/nodesPerZone/`.
</blockquote></details>

Done.

---

*[gcloud.go, line 134 at r3](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L0Glwp3UvXRoZsKoYx9-r3-134:-L1xV-HBkApKPHxlBitd:b-896fix) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c865d2a28b7f8411fb82df35404681838a48b9bb/gcloud.go#L134)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

This loop's logic deserves a comment.
</blockquote></details>

Done.

---

*[gcloud.go, line 136 at r3](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L0Glwp3UvXRoZsKoYx9-r3-136:-L1xBTVlw6UmdZcxqCsn:b-896fix) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c865d2a28b7f8411fb82df35404681838a48b9bb/gcloud.go#L136)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

you can keep the common args outside the loop as long as you use a new per-iteration arg. eg:
```
args := []string{ ....}
for .... {
  zoneArgs = append(args, newArgs...)
  run(gcloud, zoneArgs)
}
```
</blockquote></details>

Done.

---

*[gcloud.go, line 208 at r3](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L1wMbGyvpXgV1YZygoD:-L1x9sIxSd_wajf4f0JM:b-896fix) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c865d2a28b7f8411fb82df35404681838a48b9bb/gcloud.go#L208)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This is silently swallowed. You could use an `errgroup`.
</blockquote></details>

Done.

---

*[main.go, line 310 at r3](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19#-L1wMpoY-aFcImv0OAD9:-L1x9zZ6E1LHnbdNzSHC:b-896fix) ([raw file](https://github.com/cockroachlabs/roachprod/blob/c865d2a28b7f8411fb82df35404681838a48b9bb/main.go#L310)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

No dot here and below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachlabs/roachprod/19)*
<!-- Sent from Reviewable.io -->
